### PR TITLE
ledger api server: support auth0 usernames

### PR DESCRIPTION
**Pull request by meiersi-da**
### Pull Request Checklist

Fixes https://github.com/digital-asset/daml/issues/12032

FYI: @miklos-da and @digital-asset/kv-committer team, in case this conflicts with ongoing PRs.

See the commentary on `IdString.UserId` for the detailed rationale behind the user-ids being allowed.

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [x] Normal production system change, include purpose of change in description
- [x] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/IdString.scala [hunk 1]
```diff
@@ -337,17 +341,25 @@ private[data] final class IdStringImpl extends IdString {
...
-    *  Thus 'john.doe1' is a valid user-name, while 'john..doe1', 'john.-doe1', and '-john.doe' are not.
+  /** Identifiers for applications as used in custom tokens and requests.
+    * They used to be equal to [[LedgerString]], but were made to also allow '@' and '|'
+    * when introducing participant users to ensure that they are a superset of [[UserId]].
+    */
+  override type ApplicationId = String
+  override val ApplicationId: ConcatenableStringModule[LedgerString, HexString] =
+    new ConcatenableMatchingStringModule("Application ID", "._:-#/@| ", 255)
+
+  /** Identifiers for participant node users are non-empty strings that consist of ASCII alphanumeric characters and
```
Comment (cocreature) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/IdString.scala:352]: dunno how precise you want to be here in the docs but you’re missing the length limit of 62 characters.

Comment (cocreature) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/IdString.scala:352]: maybe a good idea to add a link to the auth0 restrictions so the restrictions feel a bit less arbitrary?

File: ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala [hunk 2]
```diff
@@ -294,11 +294,6 @@ object domain {
 
   sealed trait ApplicationIdTag
 
-  /** Identifiers for applications connecting to the Ledger API.
```
Comment (cocreature) [inline ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala:297]: very happy to see that go :+1:

**Review by cocreature — APPROVED** [review]
Nice! All looks good to me but I suggest waiting for someone from the participant team to review that side in detail.

File: ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala (hunk 2 above)
Comment (meiersi-da) [reply ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala:297]: you're welcome :)

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/IdString.scala (hunk 1 above)
Comment (meiersi-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/IdString.scala:352]: Good point. Reading https://auth0.com/docs/authenticate/database-connections/require-username#allowed-characters, I found that Auth0 is also quite a bit more liberal in what it allows.

Comment (meiersi-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/IdString.scala:352]: I'll adjust accordingly.

Comment (meiersi-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/IdString.scala:352]: fixed

**Comment by mziolekda** [comment]
That was helluva boilerplate there. Well done!

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/api/WriteServiceWithDeduplicationSupport.scala [hunk 3]
```diff
@@ -17,13 +16,7 @@ import com.daml.ledger.participant.state.kvutils.deduplication.{
   CompletionBasedDeduplicationPeriodConverter,
   DeduplicationPeriodSupport,
 }
-import com.daml.ledger.participant.state.v2.{
```
[the remark this answers is not in the document]
Comment (meiersi-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/api/WriteServiceWithDeduplicationSupport.scala:20]: Apologies. I can't remember doing that explicitly. I've used "Optimize Imports" extensively to get rid of the "unused import" warnings. Could that have don this as well?

**Pull request merged**